wdns/str_to_rdata_ubuf.c

```diff
@@ -263,70 +233,116 @@ str_to_svcparam(ubuf *u, char *keyval)
 		break;
 
 	/*
-	 * The wire format for IP hints is a sequence of IP addresses in
-	 * network byte order. An empty list of addresses is invalid.
+	 * The wire format for IPv{4,6} hints is a sequence of IP addresses in
+	 * network byte order. The list must not be empty.
 	 */
 	case spr_ipv4hint:
-		tok = strtok_r(NULL, ",", &endp);
+	case spr_ipv6hint: {
+		unsigned char buf[16];
+		int af = (key == spr_ipv4hint ? AF_INET : AF_INET6);
+		size_t sz = (key == spr_ipv4hint ? 4 : 16);
+
+		tok = strtok_r(val, ",", &endp);
 		if (tok == NULL || *tok == '\0') {
 			return (wdns_res_parse_error);
 		}
 
 		do {
-			if (inet_pton(AF_INET, tok, v4buf) != 1) {
+			if (inet_pton(af, tok, buf) != 1) {
 				return (wdns_res_parse_error);
 			}
 
-			ubuf_append(u, (uint8_t *)v4buf, sizeof (v4buf));
-			val_len += sizeof(v4buf);
+			ubuf_append(u, (uint8_t *)buf, sz);
+			val_len += sz;
 		} while ((tok = strtok_r(NULL, ",", &endp)) != NULL);
 		break;
-
-	case spr_ipv6hint:
-		tok = strtok_r(NULL, ",", &endp);
-		if (tok == NULL || *tok == '\0') {
+	}
+	/*
+	 * The wire format value for "alpn" consists of at least one "alpn-id"
+	 * prefixed by its length as a single octet, and these length-value
+	 * pairs are concatenated to form the SvcParamValue. These MUST exactly
+	 * fill the SvcParamValue; otherwise, the SvcParamValue is malformed.
+	 *
+	 * We parse arbitrary keyNNNN=val pairs in a similar way, but without
+	 * length-value paris (just one whole value).
+	 *
+	 * Note that we always wrap these two value types around double quotes
+	 * in presentation format, as escaped spaces and commas are allowed.
+	 */
+	case spr_alpn: {
+		ubuf *v = ubuf_new();
+		char *dup;
+		size_t i, len, offset;
+		uint8_t oclen;
+
+		if (rdata_from_str_string((const uint8_t *)val, v) <= 0) {
+			ubuf_destroy(&v);
 			return (wdns_res_parse_error);
 		}
 
-		do {
-			if (inet_pton(AF_INET6, tok, v6buf) != 1) {
-				return (wdns_res_parse_error);
-			}
+		dup = strdup(ubuf_cstr(v));
+		ubuf_destroy(&v);
 
-			ubuf_append(u, (uint8_t *)v6buf, sizeof (v6buf));
-			val_len += sizeof(v6buf);
-		} while ((tok = strtok_r(NULL, ",", &endp)) != NULL);
-		break;
+		len = strlen(dup);
+
+		offset = ubuf_size(u);
+		oclen = 0;
+		ubuf_append(u, &oclen, sizeof (uint8_t));
 
-	default:
 		/*
-		 * Parse an arbitrary keyNNNN=val pair.
-		 *
-		 * The rfc doesn't specify whether these types of keys can have
-		 * multiple values or not, so for now we treat it as a single
-		 * string (that could contain multiple comma separated values).
-		*/
-		tok = strtok_r(NULL, " ", &endp);
-		if (tok == NULL || *tok == '\0') {
-			return (wdns_res_parse_error);
-		}
+		 * Parse the resulting string for escaped commas and
+		 * backslashes.
+		 */
+		for (i = 0; i < len; i++) {
+			char c = dup[i];
 
-		val_len = ubuf_size(u);
+			if (c == '\\') {
+				/* shouldn't happen, but just to be safe.. */
+				if (++i >= len) {
+					return (wdns_res_parse_error);
+				}
+				c = dup[i];
```

Comment:
Low priority, but consider returning a parse error if character `i` is not `\` or `,` -- at this comma-separated list parsing layer, those are the only two permitted escaped characters.